ar_microcontrollers/AR4_teensy/AR4_teensy.ino

```diff
@@ -489,12 +517,17 @@ void stateTRAJ() {
         // update host with joint positions
         String msg = String("JP") + JointPosToString(curJointPos);
         Serial.println(msg);
+      } else if (function == "RE") {
```

Comment:
Fair point, I agree the current way to reset the estop is not very comfortable to use.

If we use automatic resetting (or "run only if estop not pressed") through the "MT" command, the problem is that it is published by the control node at the [configured update frequency](https://github.com/ycheng517/ar4_ros_driver/blob/eb87c547b92b525268faf3495a6102822e4e948d/ar_hardware_interface/config/controllers.yaml#L3) (100Hz), which would reset the Estop as soon as the user releases the estop push button. Since the button (at least the one [available in the robot kits](https://www.anninrobotics.com/robot-kits)) is a momentary switch, the robot would resume the potentially error prone movement after the user lets go the button. Conventionally for industrial robots, the Estop once set, stays in that state unless its twisted and pulled. This gives the user the changes to ascertain no problems exists before resuming the robot operations.

Potential solutions I can think of is to add a ros service that sends the "RE" command to the robot to reset the estop, which the user could be guided to call when they want to do so, without having to restart anything.

Alternatively, the estop reset can be something that is run when the robot is initialized using the "ST" command or immediately after calibration. That way if the user presses the estop, they would require shutting down the ros control node and restarting it to reset the estop.

What do you think?

Reply:
I see, yeah I like the convention to give user the change to ascertain that no problems exist before resuming operations. I think we could surface the e-stop to the hardware interface level through the teensy driver (i.e. send the teensy driver an `ES` reply instead of the typical reply for commands like `MT`) and then the hardware interface can return error (which I think makes the controller deactivated?). When the e-stop is cleared and we're ready to resume, we activate the driver via `ros2 control set_controller_state active` which calls the `RE` command to reset the e-stop. This way on the ROS side we stay in an established error handling paradigm, and the user is aware that the software is in an e-stop state. Below are example code snippets illustrating the concept: 
---

returning error during read:
```
hardware_interface::return_type MyHardwareInterface::read()
{
    if (e_stop_condition) {
        return hardware_interface::return_type::ERROR;
    }
    // Normal read operation
    return hardware_interface::return_type::OK;
}
```

resetting e-stop on activate:
```
controller_interface::return_type MyController::on_activate(
    const rclcpp_lifecycle::State & /*previous_state*/)
{
  bool sucess = driver_.reset_estop();
    if (!success) {
        RCLCPP_ERROR(get_node()->get_logger(), "Cannot activate. Hardware in E-stop state.");
        return controller_interface::return_type::ERROR;
    }

    RCLCPP_INFO(get_node()->get_logger(), "Controller activated.");
    return controller_interface::return_type::OK;
}
```

Reply:
I like your idea! It gives more streamlined control to the user and the ros2 controller is kept informed of the state of the hardware through the interface. I'll add in those changes and run some tests to make sure its functional :robot:

Reply:
great, btw I have a sizable change coming up in https://github.com/ycheng517/ar4_ros_driver/pull/28 that makes the robot arm move much faster amongst other improvements. Feel free to try it out!

Reply:
Implemented the feature as we discussed, with some changes:
- The teensy driver receives the ES status as a response to any MT commands, as a result the hardware interface deactivates when estop is detected within its `write()` method (not `read()`)
- Upon returning `hardware_interface::return_type::ERROR` from the `write()` method, the hardware interface is disabled along with the joint_state controller and broadcaster as well (see image and video below). To reset the estop, I have to set all 3 to active using
  ```bash
  ros2 control set_hardware_component_state mk3 active
  ros2 control set_controller_state joint_trajectory_controller active
  ros2 control set_controller_state joint_state_broadcaster active
  ```
- When estop is pressed during a motion, and then reset by activating the interfaces using instructions above, the robot jerks, indicative of trying to resume its acceleration/deceleration profiles. To avoid this, all joints accelerations and step intervals are reset using `stepperJoints[i].setCurrentPosition(stepperJoints[i].currentPosition())`, since that is the only `public` method available to do that

With these changes though, the estop functionality seems to work well IMO. I'd like to know if it works for you too

Before estop:
![Screenshot from 2024-11-25 00-20-58](https://github.com/user-attachments/assets/9fe751b0-cf24-4dab-8e62-ef5571cc2d34)

After estop:
![Screenshot from 2024-11-25 00-21-26](https://github.com/user-attachments/assets/34ecc85b-5a66-40eb-988d-8358504a370f)

video: estop pressed during motion, interfaces are disabled, then estop reset by activating the interfaces back up, which enables using the driver again
https://github.com/user-attachments/assets/57b3ed7d-e9e0-4215-83bc-5e6afb42d0b3

Reply:
great problem solving to get this working well!